Change SCP's Signature definition to be a fixed-size type

MichaelKim20 opened the pull request:
I changed `Signature` to fixed length array.

``` C++
using Signature = xdr::opaque_vec<64>;
```
I changed it as follows.
``` C++
using Signature = xdr::opaque_array<64>;
```

`Signature` is the data structure used by SCP.
This is declared vector in SCP because of compatibility between versions.
SCP has several versions. In some of the previous versions, Signature was 32 Byte and is now 64 Byte.

For this reason, I think there is no problem if we change to 64Byte fixed length.

And I tried to compile it after changing it in `stellar-core`.
There were some errors in the test codes, but there were no problems with the rest.
[

https://github.com/stellar/stellar-core/blob/69c75632ae903191a82c7b2b010cce4022a60870/src/crypto/SecretKey.cpp#L128

https://github.com/stellar/stellar-core/blob/69c75632ae903191a82c7b2b010cce4022a60870/src/transactions/test/TxEnvelopeTests.cpp#L83-L101

https://github.com/stellar/stellar-core/blob/69c75632ae903191a82c7b2b010cce4022a60870/src/herder/test/HerderTests.cpp#L1262-L1266
]
Relates to #744

AndrejMitrovic commented on source/agora/utils/SCPPrettyPrinter.d:
```diff
@@ -459,7 +459,7 @@ GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000)], enrolls: [{ utxo: 0x0000...e2
     assert(MissingSig == format("%s", scpPrettify(&env)),
```
In this module there is a warning:

> // be warned: toVec() will point to stack-allocated memory

This warning should now be removed. Actually there is a call to `toVec()` in `env.signature = sig[].toVec();`, so the `toVec()` call should be removed too. It should be just `env.signature = pair.secret.sign(msg[]);`

Geod24 replied:
Looks like it's been addressed ? I don't see the warning in this module.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -761,26 +761,23 @@ private struct SCPEnvelopeHash
     assert(getEnvHash().hashFull() == Hash.fromString(
-        "0xd3dc2318365e55ea3a62b403fcbe22d447402741a5151a703326dbf852350dcd0" ~
-        "e020a762ae12a871473aed80d82f51c1cd3942c0b2360c2d609279a2867fe68"),
+        "0xfd2ea2b85d2a315a9817e6661bc3c4378637de37649d2fdb9ca82d6e4172e9e46" ~
```
The reason why this changed is because the signature can no longer be empty when it's a fixed-length array. It wasn't obvious at first.

MichaelKim20 replied:
I think the length of 'xdr:::opaque_vec' is recorded in front  when it is serialized.

AndrejMitrovic replied:
It's not about serialization, just about hashing. In this specific case the unittest was testing what happens when the signature is empty, because a dynamic array can be empty. But now that the signature is a fixed length array, it's not possible to test for an "empty" signature.

Anyway your code is correct, I just made a comment in case anyone else is confused why the hash changed.

MichaelKim20 replied:
Yeah, I checked. Thank you.

AndrejMitrovic commented on source/scpp/src/crypto/SecretKey.cpp:
```diff
@@ -91,7 +93,7 @@ SecretKey::sign(ByteSlice const& bin) const
     assert(mKeyType == PUBLIC_KEY_TYPE_ED25519);
 
-    Signature out(crypto_sign_BYTES, 0);
+    Signature out;
```
How about using `Signature out(0);`. Does it work?

MichaelKim20 replied:
I will try it.

MichaelKim20 replied:
I tried this, and this is what I got.
```
SecretKey.cpp:96:15: error: no matching constructor for initialization of 'stellar::Signature' (aka 'opaque_array<64>')
    Signature out(0);
```

AndrejMitrovic commented on source/scpp/src/xdr/Stellar-types.x:
```diff
@@ -52,7 +52,7 @@ case SIGNER_KEY_TYPE_HASH_X:
 };
 
 // variable size as the size depends on the signature scheme used
```
This comment can be replaced:

```
// fixed size as we use a 64-byte Signature in Agora
```

AndrejMitrovic reviewed (approved):
Just small comments, it LGTM.

Geod24 commented on source/scpp/extra/DSizeChecks.cpp:
```diff
@@ -46,7 +46,7 @@ CPPSIZEOF(CryptoKeyType)
 CPPSIZEOF(SignerKeyType)
 CPPSIZEOF(PublicKey)
-CPPSIZEOF(Signature)
+// Signature is removed because it is the same as Hash.
```
Does the test cause problem ?

MichaelKim20 replied:
I could see the following.
```
Build failed: /agora-node/agora/source/scpp/extra/DSizeChecks.cpp:49:1: error: redefinition of 'cppSizeOf'
CPPSIZEOF(Signature)
^
```

MichaelKim20 replied:
I changed the pointer of SCPEnvelope to non-pointer.

MichaelKim20 replied:
Could you check how it is?

MichaelKim20 replied:
There was no warning. But I think it's better to change the same as other structs

AndrejMitrovic replied:
Could we add something like a static assert in the C++ code?

```C++
#include <type_traits> 

static_assert(std::is_same<Signature, Hash>::value, "Signature and Hash must be the same type");
```

Geod24 reviewed (commented):
Just a couple questions, looks good

MichaelKim20 commented:
@AndrejMitrovic After your approval, I have modified the code, so please review it again.

AndrejMitrovic commented:
What did you change?

AndrejMitrovic commented on source/agora/utils/SCPPrettyPrinter.d:
```diff
@@ -317,30 +301,12 @@ private struct SCPStatementFmt
 private struct SCPEnvelopeFmt
 {
-    /// Pointer to SCPEnvelope
-    private const SCPEnvelope* envelope;
```
The reason why this was a pointer was to avoid memory copies.

This is even more important now that the signature is a static array.

MichaelKim20 replied:
@Geod24, @AndrejMitrovic  Okay. I'll remove the last commit.

linked0 reviewed (approved):
I checked all the code and LGTM.

Geod24 reviewed (approved):
Just include https://github.com/bpfkorea/agora/pull/806/files#r425007595 and LGTM

MichaelKim20 committed:
Change SCP's Signature definition to be a fixed-size type

Change the hash test of SCPEnvelope
Change the size checking of SCPEnvelope
Change the initialization of Signature in SecretKey::sign

MichaelKim20 committed:
Modify to use directly without creating a new `Signature` instance

MichaelKim20 commented:
> Just include https://github.com/bpfkorea/agora/pull/806/files#r425007595 and LGTM

It's just been completed
